libraries/ms-common/src/stm32f0xx/soft_timer.c

```diff
@@ -0,0 +1,211 @@
+                                          .TIM_ClockDivision = TIM_CKD_DIV1 };
+
+  TIM_TimeBaseInit(TIM2, &init_struct);
+
+  // Reset the counter to 0.
+  TIM_SetCounter(TIM2, 0);
+
+  // Disable all the timers and forget the last running timer.
+  s_last_update_count = 0;
+  s_active_timer_id = SOFT_TIMER_MAX_TIMERS;
+  for (uint32_t i = 0; i < SOFT_TIMER_MAX_TIMERS; i++) {
+    s_soft_timer_array[i].inuse = false;
+  }
+
+  // Enable the interrupts. TIM2 uses IRQ channel 15.
+  stm32f0xx_interrupt_nvic_enable(15, INTERRUPT_PRIORITY_NORMAL);
+  TIM_ITConfig(TIM2, TIM_IT_CC1, ENABLE);
+
+  // Start the timer.
+  TIM_Cmd(TIM2, ENABLE);
+}
+
+StatusCode soft_timer_start(uint32_t duration_us, SoftTimerCallback callback, void *context,
+                            SoftTimerID *timer_id) {
+  // Enable a critical section.
+  bool critical = critical_section_start();
+  for (uint32_t i = 0; i < SOFT_TIMER_MAX_TIMERS; i++) {
+    if (!s_soft_timer_array[i].inuse) {
+      // Look for an empty timer.
+
+      // Get the time this was called.
+      uint32_t update_elapsed_time = prv_soft_timer_update_period();
+      if (s_active_timer_id >= SOFT_TIMER_MAX_TIMERS) {
+        // New timer will be the only timer. Start it.
+        s_active_timer_id = i;
+        prv_start_timer(duration_us);
+      } else if (duration_us <
+                 s_soft_timer_array[s_active_timer_id].duration - update_elapsed_time) {
+        // New timer will run be before the active timer. Update and start it. Use as close to now
+        // as possible for the update.
+        prv_soft_timer_update(prv_soft_timer_update_period());
+        s_active_timer_id = i;
+        prv_start_timer(duration_us);
+      } else {
+        // Update the current time so as to not impact this timer's duration.
+        prv_soft_timer_update(update_elapsed_time);
+        prv_start_timer(s_soft_timer_array[s_active_timer_id].duration -
+                        prv_soft_timer_update_period());
+      }
+      // Otherwise, the new timer is longest ignore it until later.
+
+      // Actually populate the timer. Do this after the previous steps as this is a critical section
+      // and this timer shouldn't be updated in prv_soft_timer_update.
+      s_soft_timer_array[i].duration = duration_us;
+      s_soft_timer_array[i].callback = callback;
+      s_soft_timer_array[i].context = context;
+      s_soft_timer_array[i].inuse = true;
+      *timer_id = i;
+      critical_section_end(critical);
+      return STATUS_CODE_OK;
+    }
+  }
+
+  // Out of timers.
+  critical_section_end(critical);
+  return status_msg(STATUS_CODE_RESOURCE_EXHAUSTED, "Out of software timers.");
+}
+
+bool soft_timer_inuse(void) {
+  if (s_active_timer_id != SOFT_TIMER_MAX_TIMERS) {
+    return true;
+  }
+  return false;
+}
+
+// TIM2 Interrupt handler as defined in stm32f0xx.h
+void TIM2_IRQHandler(void) {
+  if (TIM_GetITStatus(TIM2, TIM_IT_CC1) != RESET) {
```

Comment:
I'm a little worried about the length of this ISR. It's a shame we don't have a cycle counter on this chip. :P

We can probably estimate it by using a GPIO to determine the time it takes to run. Since we know the clock frequency, we can estimate the number of cycles over the time the GPIO was active.